src/rvsdg/egglog_optimizer/subst.rs

```diff
@@ -42,12 +44,144 @@ fn subst_all_rule(btype: String) -> String {
+        (rule ((can-subst-Body-beneath above from to)
+               (= new-from (Project i from)))
+              ((can-subst-Operand-beneath above new-from (Project i to)))
+              :ruleset subst)
+
+        ;; Learn can-subst-body-beneath
+        (rule ((can-subst-Expr-beneath above from to)
+               (= new-from (PureOp from)))
+              ((can-subst-Body-beneath above new-from (PureOp to)))
+              :ruleset subst)
+        ;; Propagates up same context (Gamma: pred & inputs, Theta: inputs)
+        ;; rtjoa: Is it sound to propagate up outputs if we renumber args?
+        (rule ((can-subst-Operand-beneath above from to)
+               (= new-from (Gamma from inputs outputs)))
+              ((can-subst-Body-beneath above new-from (Gamma to inputs outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= new-from (Gamma pred from outputs)))
+              ((can-subst-Body-beneath above new-from (Gamma pred to outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= new-from (Theta pred from outputs)))
+              ((can-subst-Body-beneath above new-from (Theta pred to outputs)))
+              :ruleset subst)
+        (rule ((can-subst-VecOperand-beneath above from to)
+               (= new-from (OperandGroup from)))
+              ((can-subst-Body-beneath above new-from (OperandGroup to)))
+              :ruleset subst)
+        "
+    .into()];
+
+    // Learn can-subst-expr-beneath
+    res.push(
+        "
+      (rule ((can-subst-VecOperand-beneath above from to)
+              (= new-from (Call ty f from n-outs)))
+             ((can-subst-Expr-beneath above new-from (Call ty f to n-outs)))
+            :ruleset subst)"
+            .into(),
+    );
+    for bril_op in BRIL_OPS {
+        let op = bril_op.op;
+
+        match bril_op.input_types.as_ref() {
+            [Some(_), Some(_)] => res.push(format!(
+                "
+              (rule ((can-subst-Operand-beneath above from to)
+                      (= new-from ({op} type from e2)))
+                     ((can-subst-Expr-beneath above new-from ({op} type to e2)))
+                    :ruleset subst)
+              (rule ((can-subst-Operand-beneath above from to)
+                      (= new-from ({op} type e1 from)))
+                     ((can-subst-Expr-beneath above new-from ({op} type e1 to)))
+                    :ruleset subst)
+                     ",
+            )),
+            _ => unimplemented!(),
+        };
+    }
+
+    // Learn can-subst-{VecOperand, VecVecOperand}-beneath
+    for (vectype, ctor, eltype) in &[
+        ("VecOperand", "VO", "Operand"),
+        ("VecVecOperand", "VVO", "VecOperand"),
+    ] {
+        // Reify vec-get to be able to be able to match on vec contents
+        res.push(format!(
+            "
+          (function {vectype}-get ({vectype} i64) {eltype})
```

Comment:
Can't decide if this is a hack or nice solution XD
Let's call it a nice solution for now. These vectors are going to be very slow though with all the vec-set going on.

Reply:
Yeah, I think it's less of a hack though than O(n^2) subst in vectors with many vec-sets for `subst`, since this will stay linear at least.

Agreed, once we tighten up vecs then most code using them will likely change.